algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py

```diff
@@ -212,122 +210,50 @@ def forward(self, inputs, input_paddings):
-    running_var = (
-        self.running_var
-        if not self.training or self.track_running_stats else None)
-
-    # Don't sync batchnorm stats in inference mode (model.eval()).
-    need_sync = (
-        bn_training and self.training and USE_PYTORCH_DDP and input.is_cuda)
-    if need_sync:
-      process_group = torch.distributed.group.WORLD
-      if self.process_group:
-        process_group = self.process_group
-      world_size = torch.distributed.get_world_size(process_group)
-      need_sync = world_size > 1
-
-    # fallback to framework BN when synchronization is not necessary
-    if not need_sync:
-      return F.batch_norm(
-          input,
-          running_mean,
-          running_var,
-          1 + self.weight,
-          self.bias,
-          bn_training,
-          exponential_average_factor,
-          self.eps,
-      )
-    else:
-      assert bn_training
-      return sync_batch_norm.apply(
-          input,
-          1 + self.weight,
-          self.bias,
-          running_mean,
-          running_var,
-          self.eps,
-          exponential_average_factor,
-          process_group,
-          world_size,
-      )
-
-
 class BatchNorm(nn.Module):
 
   def __init__(self, dim, batch_norm_momentum, batch_norm_epsilon):
     super().__init__()
-    self.bn = CustomBatchNorm1d(
-        num_features=dim,
-        momentum=1 - batch_norm_momentum,
-        eps=batch_norm_epsilon)
-
-  def forward(self, inputs, input_paddings=None):
-    # inputs: ...D
-    #padding: ...
-    *b, d = inputs.shape
-    reduce_size = functools.reduce(lambda x, y: x * y, b)
-
-    inputs = inputs.reshape(reduce_size, d)
-    if input_paddings is None:
-      input_paddings = torch.zeros(size=b, device=inputs.device)
-    input_paddings = input_paddings.reshape(reduce_size)
-    bn_inp = self.bn(inputs[input_paddings == 0])
-    output = torch.zeros(reduce_size, d, device=inputs.device)
-    output[input_paddings == 0] = bn_inp
+    running_mean = torch.zeros(dim)
+    running_var = torch.ones(dim)
+    self.register_buffer('running_mean', running_mean)
+    self.register_buffer('running_var', running_var)
+    self.weight = nn.Parameter(torch.zeros(dim))
+    self.bias = nn.Parameter(torch.zeros(dim))
+
+    self.momentum = batch_norm_momentum
+    self.epsilon = batch_norm_epsilon
+    self.dim = dim
 
-    return output.reshape(*b, d)
+  def forward(self, inputs, input_paddings):
+    #inputs: NHD
+    #padding: NH
+    mask = 1 - input_paddings[:, :, None]
+    if self.training:
+      count = mask.sum()
+      masked_inp = inputs.masked_fill(mask == 0, 0)
+      sum_ = (masked_inp).sum(dim=(0, 1))
+      if USE_PYTORCH_DDP:
+        sum_ = dist_nn.all_reduce(sum_)
+        count = dist_nn.all_reduce(count)
+      mean = sum_ / count
+
+      sum_ = (torch.square(masked_inp - mean) * mask).sum(dim=(0, 1))
+      if USE_PYTORCH_DDP:
+        sum_ = dist_nn.all_reduce(sum_)
```

Comment:
seems like with this change torch batchnorm is closer to custom jax ones , do conformer steps match without this change ? or is this change required for conformer steps to match ?

Reply:
I did this change for the deepspeech model; the earlier batchnorm implementation is also technically identical to this but made use of the in-built syncbatchnorm for synchronizing the batch stats.